src/sqlfluff/core/parser/lexer.py

```diff
@@ -561,27 +633,16 @@ def elements_to_segments(
             # Generate placeholders for any source-only slices that *follow*
             # the last element. This happens, for example, if a Jinja templated
             # file ends with "{% endif %}", and there's no trailing newline.
-            if idx == len(elements) - 1:
-                so_slices = [
-                    so
-                    for so in source_only_slices
-                    if so.source_idx >= source_slice.stop
-                ]
-                for so_slice in so_slices:
-                    segment_buffer.append(
-                        TemplateSegment(
-                            pos_marker=PositionMarker(
-                                slice(so_slice.source_idx, so_slice.end_source_idx()),
-                                slice(
-                                    element.template_slice.stop,
-                                    element.template_slice.stop,
-                                ),
-                                templated_file,
-                            ),
-                            source_str=so_slice.raw,
-                            block_type=so_slice.slice_type,
-                        )
-                    )
+            if last_source_slice and idx == len(elements) - 1:
+                template_segments, _ = _generate_placeholder_segments(
+                    slice(source_slice.stop, len(templated_file.source_str)),
+                    last_source_slice,
+                    slice(element.template_slice.stop, element.template_slice.stop),
+                    source_only_slices,
+                    templated_file,
+                    add_indents,
+                )
+                segment_buffer.extend(template_segments)
```

Comment:
This is the main _behaviour_ change here. The rest is test cases or reorganisation. Previously it had it's own logic. Now it reuses the logic from above.